File: 프로그래머스/2/155651. 호텔 대실/호텔 대실.py
def solution(book_time):
    book_time = sorted(list(map(change_min, book_time)), key=lambda x: (x[0], x[1]))
    
    print(book_time)
    answer = [book_time[0][1]]
    
    for i in range(1, len(book_time)):
        index = find_room(answer, book_time[i][0])
        if index == -1:
            answer.append(book_time[i][1])
        else:
            answer[index] = book_time[i][1]
    
    return len(answer)
                


def change_min(time):
    h1, m1 = map(int, time[0].split(":"))
    h2, m2 = map(int, time[1].split(":"))
    
    
    return [h1*60+m1, h2*60+m2]


def find_room(answer, time):
    for i in range(len(answer)):
        if answer[i] + 10 <= time:
            return i
    return -1
                

File: 프로그래머스/2/155651. 호텔 대실/호텔 대실.java
import java.util.*;
import java.util.stream.*;

class Solution {
    public int solution(String[][] book_time) {
        List<int[]> bookTimes = Arrays.stream(book_time)
            .map(times -> {
                String[] start = times[0].split(":");
                int startTime = Integer.parseInt(start[0]) * 60 + Integer.parseInt(start[1]);
                
                String[] end = times[1].split(":");
                int endTime = Integer.parseInt(end[0]) * 60 + Integer.parseInt(end[1]) + 10;
                
                return new int[]{startTime, endTime};  
            })
            .sorted(Comparator.comparingInt((int[] x) -> x[0]))
            .collect(Collectors.toList());
        
        
        List<Integer> room = new ArrayList<>();
        
        for(int[] bookTime : bookTimes) {
            int start = bookTime[0];
            int end = bookTime[1];
            boolean isAvailable = false;
            for (int i = 0; i < room.size(); i++) {
                if (start >= room.get(i)) {
                    room.set(i, end);
                    isAvailable = true;
                    break;
                }
            }
            
            if (!isAvailable) {
                room.add(end);
            }
        }
        
        
        
        return room.size();
        
    }
}